File: src/main/java/ru/arc/util/ParticleManager.java
package ru.arc.util;

import ru.arc.ARC;
import com.destroystokyo.paper.ParticleBuilder;
import lombok.extern.slf4j.Slf4j;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.*;
import java.util.concurrent.ConcurrentLinkedDeque;

@Slf4j
public class ParticleManager {

    private static final Deque<ParticleBuilder> buildersQueue = new ConcurrentLinkedDeque<>();
    private static final Deque<ParticleBuilder> syncBuildersQueue = new ConcurrentLinkedDeque<>();

    private static BukkitTask task, syncTask;

    public static void setupParticleManager() {
        if (task != null && !task.isCancelled()) task.cancel();
        if (syncTask != null && !syncTask.isCancelled()) syncTask.cancel();

        task = new BukkitRunnable() {
            @Override
            public void run() {
                while (!buildersQueue.isEmpty()) buildersQueue.poll().spawn();
            }
        }.runTaskTimerAsynchronously(ARC.plugin, 0L, 1L);

        syncTask = new BukkitRunnable() {
            @Override
            public void run() {
                int count = 0;
                while (!syncBuildersQueue.isEmpty()) {
                    syncBuildersQueue.poll().spawn();
                    count++;
                    if (count > 200) {
                        log.warn("Too many particles to show in one tick. Size: {}", syncBuildersQueue.size());
                        break;
                    }
                }
            }
        }.runTaskTimer(ARC.plugin, 0L, 1L);
    }

    public static void queue(ParticleBuilder builder) {
        boolean res = buildersQueue.offer(builder);
        if (!res) log.warn("Failed to queue particle builder: {}", builder);
    }
}

File: src/main/kotlin/ru/arc/util/ParticleManager.kt
package ru.arc.util

import com.destroystokyo.paper.ParticleBuilder
import ru.arc.core.ScheduledTask
import ru.arc.core.repeating
import ru.arc.core.ticks
import ru.arc.util.Logging.warn
import java.util.concurrent.ConcurrentLinkedDeque

/**
 * Batches particle spawning on the server thread.
 *
 * Particle builders eventually call the Bukkit world API, so both legacy queue
 * entry points are drained by the same synchronous task.
 */
object ParticleManager {

    private val buildersQueue = ConcurrentLinkedDeque<ParticleBuilder>()
    private val syncBuildersQueue = ConcurrentLinkedDeque<ParticleBuilder>()

    private var task: ScheduledTask? = null

    @JvmStatic
    fun setupParticleManager() {
        stopTasks()
        task =
            repeating(period = 1.ticks, delay = 0.ticks) {
                var count = 0
                while (count < MAX_PARTICLES_PER_TICK) {
                    val builder = syncBuildersQueue.poll() ?: buildersQueue.poll() ?: break
                    builder.spawn()
                    count++
                }
                val remaining = syncBuildersQueue.size + buildersQueue.size
                if (remaining > 0) {
                    warn("Particle queue rate-limited; {} builders remain", remaining)
                }
            }
    }

    @JvmStatic
    fun stopTasks() {
        task?.cancel()
        task = null
        buildersQueue.clear()
        syncBuildersQueue.clear()
    }

    @JvmStatic
    fun queue(builder: ParticleBuilder) {
        if (!buildersQueue.offer(builder)) {
            warn("Failed to queue particle builder: {}", builder)
        }
    }

    @JvmStatic
    fun queueSync(builder: ParticleBuilder) {
        if (!syncBuildersQueue.offer(builder)) {
            warn("Failed to queue sync particle builder: {}", builder)
        }
    }

    private const val MAX_PARTICLES_PER_TICK = 200
}
